File: app/src/main/java/br/com/intelbras/api/util/PokemonListDeserializer.java
package br.com.intelbras.api.util;

import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

import br.com.intelbras.model.Pokemon;


public class PokemonListDeserializer implements JsonDeserializer<List<Pokemon>> {
    @Override
    public List<Pokemon> deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        List<Pokemon> items = new ArrayList<>();

        final JsonObject jsonObject = json.getAsJsonObject();
        final JsonArray itemsJsonArray = jsonObject.get("results").getAsJsonArray();

        for (JsonElement itemsJsonElement : itemsJsonArray) {
            Pokemon p = new Pokemon();
            final JsonObject itemJsonObject = itemsJsonElement.getAsJsonObject();
            p.setName(itemJsonObject.get("name").getAsString());
            p.setUrl(itemJsonObject.get("url").getAsString());

            items.add(p);
        }

        return items;
    }
}

File: app/src/main/java/br/com/intelbras/api/util/PokemonListDeserializer.kt
package br.com.intelbras.api.util


import br.com.intelbras.model.Pokemon
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import java.lang.reflect.Type
import java.util.*

class PokemonListDeserializer : JsonDeserializer<List<Pokemon>> {
    @Throws(JsonParseException::class)
    override fun deserialize(
        json: JsonElement,
        typeOfT: Type,
        context: JsonDeserializationContext
    ): List<Pokemon> {
        val items: MutableList<Pokemon> = ArrayList()
        val jsonObject = json.asJsonObject
        val itemsJsonArray = jsonObject["results"].asJsonArray
        for (itemsJsonElement in itemsJsonArray) {
            val p = Pokemon()
            val itemJsonObject = itemsJsonElement.asJsonObject
            p.name = itemJsonObject["name"].asString
            p.url = itemJsonObject["url"].asString
            items.add(p)
        }
        return items
    }
}